### ISSUE TITLE
MarketLatentStateLearner cannot observe the market: its latents are frozen parameters

### ISSUE THREAD

**Issue by magilliam27** [magilliam27/MCI-GRU#198]
**Mode:** AFK. Fully specified; an agent can resolve this alone.

**Owned paths:**

- `mci_gru/models/latent.py`
- `mci_gru/models/trunk.py`
- `mci_gru/models/factory.py`
- `mci_gru/config.py`
- `tests/test_market_latent_state.py` (new)
- `docs/CONFIGURATION_GUIDE.md` (one row in the model table)

No other path is touched, and **no defaults change**: `configs/config.yaml`, the presets under `configs/experiment/`, and `docs/DEFAULT_EXPERIMENT_RECIPE.md` are out of scope.

## Summary

`MarketLatentStateLearner` produces the B1 and B2 streams that `docs/ARCHITECTURE.md` calls "market latent states", and that the model is named for (the multi-head **C**ross-attention in MCI-GRU). **It cannot observe the market.**

`R1` and `R2` are plain `nn.Parameter` tensors, and `forward` receives only the per-stock vectors:

```python
# mci_gru/models/latent.py
self.R1 = nn.Parameter(torch.randn(num_latent_states, feature_dim) * latent_init_scale)
self.R2 = nn.Parameter(torch.randn(num_latent_states, feature_dim) * latent_init_scale)

def forward(self, A1: torch.Tensor, A2: torch.Tensor) -> tuple[torch.Tensor, torch.Tensor]:
    ...
```

So `B1 = f(A1)` and `B2 = f(A2)` pointwise, and after training `f` is fixed. Nothing about the date enters. On a day the market falls five percent and a day it rises five percent, `R1` and `R2` are byte for byte identical, and a stock whose own features are unchanged receives an identical "market latent state" on both.

What the module actually computes is a learned codebook: each stock's vector re-expressed as a soft mixture of 32 prototypes. That is a per-stock nonlinearity. It is a reasonable thing for a network to contain, but it is not a market state, and the architecture documentation describes it as one.

## Why this has consequences beyond naming

`ICLoss` centres predictions within each date, so any function of the date alone is invisible to the objective. The global regime, VIX and credit features are broadcast identically to every stock. With frozen latents there is no mechanism that turns those broadcast values into a per-stock effect, so they can only act through interactions the network must discover unaided.

That is consistent with the standing observation in `docs/ABLATION_NOTEBOOK_RESULTS_REPORT_2026-04-30.md` that regime context behaves as a secondary tuning dimension rather than a driver. This issue does not claim to explain that result, but it identifies a mechanism that would produce it.

## Evidence

Read directly from the module source, not inferred: `R1`/`R2` are `nn.Parameter`, and `forward(self, A1, A2)` takes no date, batch, mask, or cross-section argument. Both the legacy 8-Linear path and the `use_nn_multihead_attention` path use them as fixed keys and values.

Analysis in `docs/research/current/MCI_GRU_TRUNK_ARCHITECTURE_OPPORTUNITIES_2026-09-05.md`, sections 1 (Defect 2), 2, and 4.2.

## What this issue does NOT claim

**No performance claim.** Two-seed smokes could not separate any architecture variant from any other; the within-variant seed spread was roughly seven times the paired protocol's 0.0035 detectable effect (section 3.8). This is filed as a **correctness defect**: a module whose documented purpose is unachievable by its construction.

## Seams under test

Per `/mattpocock-skills:tdd`, declared before any test is written.

**Tested at:**

- `MarketLatentStateLearner.forward(...)` — the module's public boundary, including the new optional arguments.
- `mci_gru.models.create_model(...)` and `StockPredictionModel.forward(...)` — construction, checkpoint compatibility, end-to-end shape and masking.
- `ModelConfig` construction and `ModelConfig.to_dict()`.

**Explicitly not tested:** attention weights, the latent tensors' values, or any private helper.

The central specification test, which the shipped module fails and the fix must pass:

> Holding one stock's own inputs fixed and changing **other** stocks on the same date must change that stock's market-latent output.

## Proposed change

Make the latents data-dependent, config-gated, default off. Gather from the date's cross-section, then broadcast, which is the Set Transformer induced-set construction (Lee et al. 2019):

```
R_d = R + MHA(query=R, key/value=A_active_on_this_date)   # latents read the market
B   = MHA(query=A, key/value=R_d)                          # unchanged second step
```

1. Add `market_latent_mode: str = "static"`, valid `{"static", "data_dependent"}`, to `ModelConfig`, validated in `__post_init__` and serialised by `to_dict`.
2. Extend `MarketLatentStateLearner.forward` with optional `num_stocks: int | None = None` and `stock_mask: torch.Tensor | None = None`. Both default to `None`, so every existing call site keeps its current behaviour.
3. In `data_dependent` mode, build `R1_d` and `R2_d` with one gather each, excluding PIT-inactive names via `key_padding_mask`, then feed them to the existing cross-attention.
4. Route the flag through `create_model` with a legacy-safe `config.get(...)` default, and pass `num_stocks` and `stock_mask` from `StockPredictionModel.forward`.

**Cost is small here**, which is the reason this placement is preferred over putting the same construction in the cross-stock position. The latent learner works at `align_dim` (32 by default), not at `concat_size` (128), so two gather modules add roughly 8K parameters rather than the 137K the cross-stock placement measured, and 32 latents against about 110 active names is a real reduction rather than a token one.

## Acceptance criteria

- [ ] With default config, `create_model` produces a model whose `state_dict()` keys and parameter count are unchanged, and a state dict saved before this change loads with `strict=True`.
- [ ] `MarketLatentStateLearner.forward(A1, A2)` with no new arguments behaves exactly as today (bitwise identical for a fixed seed).
- [ ] In `data_dependent` mode, changing other stocks on a date changes a held-fixed stock's B1 and B2. In `static` mode it does not. This is the load-bearing test and must be mutation-checked in both directions.
- [ ] PIT-inactive names do not contribute to the gathered latents, and inactive nodes remain exactly zero at the model output.
- [ ] A date with no active names does not produce NaN.
- [ ] Finite output under `torch.autocast` on CPU bfloat16.
- [ ] `ModelConfig` rejects an invalid value and `to_dict()` round-trips the new field.
- [ ] Mutation table posted on the pull request, covering **inversions as well as deletions**.
- [ ] `ruff check .` and `ruff format --check .` clean; full suite green via `scripts/run_pytest_isolated.py`.
- [ ] `docs/TEST_REGISTRY.md` regenerated.

## Out of scope

- Any default, preset, or frozen-recipe change.
- Placing this construction in the cross-stock block position; that is an ablation variant, not a defect fix, and it is priced very differently.
- The cheaper market-gate alternative (scaling inputs by a per-date market vector). It is a different mechanism and belongs in its own issue if it is wanted.
- The cross-stock `SelfAttention` residual defect, filed separately.
- Any performance claim.

**Comment by magilliam27** [comment]
**Owned paths expanded during implementation.** Recorded here rather than only on the pull request, since a concurrent session reads the ticket.

| Path | Why |
| --- | --- |
| `docs/ARCHITECTURE.md` | The review called this a hard violation of the source-of-truth policy. Its "B1/B2: market latent state" section asserted the latents are always frozen parameters, which this change makes false, and every sibling model knob is documented there. |
| `configs/config.yaml` | Same reason as #197: `AGENTS.md` routes config through Hydra YAML, and without a row the knob needs `+model.` append syntax. Added at value `static`, which **is** the existing default, so no default changes. |
| `docs/TEST_REGISTRY.md` | Required by this issue's own acceptance criteria. |

`README.md` has the same stale description of B1/B2 and was **not** touched, being outside anything this issue owns. Worth a follow-up.

## Four review findings changed the implementation

1. **Silent config override, now refused.** In `data_dependent` mode the legacy 8-Linear cross-attention is unreachable, and the first draft simply ignored `use_nn_multihead_attention` — which `paper_faithful.yaml` pins to `false`. That is the defect family recorded in #131: a config surface whose real behaviour differs from its apparent one. `ModelConfig` now raises on the combination instead, and the module keeps the attribute honest.

2. **The bitwise-identity criterion was prose, not a test.** `test_static_mode_ignores_the_new_arguments` only compared new code against new code, so any regression in the static branch would appear on both sides. Added golden values captured by running `origin/main`'s module at `125abda`, pinning **both** static branches.

3. **No strict checkpoint load was exercised.** Added one.

4. **The mutation table found two redundant masking steps and one dead guard.** Zeroing an inactive row and excluding it as an attention key are each sufficient for content independence, so removing either survived. What distinguishes them is dilution: a zeroed row still occupies a slot in the softmax. Two new guards pin that, at the module and model seams, and the now-redundant row zeroing was deleted. The empty-date guard was also deleted after checking that torch 2.11 returns zeros rather than NaN for a fully-masked row; the no-NaN test remains as the canary, and the reliance is stated in a comment.

## Not actioned

- **`VALID_MODES` duplicated** between `latent.py` and `config.py`. Kept deliberately: `MarketLatentStateLearner` is constructible directly, as these tests do, so it needs its own guard rather than trusting every caller to come via `ModelConfig`.
- **Minor duplication** flagged in the gather (`batch` recomputed, `.view` vs `.reshape`). Cosmetic; left rather than widening the diff.
